### ISSUE TITLE
How do I create element with template on the fly

### ISSUE THREAD

**Issue by antitoxic** [skatejs/skatejs#311]
What is the intended way to create custom element dynamically while this element has template function that uses the inner html.

```js
var el = document.createElement('my-custom-el');
// template is already rendered here
el.append(innerHtmlNode)
skate.init(el)
```

The problem is that the moment I'm appending content, the template is already in there. I can't find a non-messy way to do it.

**Comment by treshugart** [comment]
That's a templating issue and Skate has no opinions about templating other than it goes as far to help facilitate multiple methods of it.

The best way to achieve exactly what you're trying to do is to use a shadow root.

I'm on the bus at the moment and will add a few examples of alternate methods when I'm at my desk.

**Comment by treshugart** [comment]
OK! I'm in.

There's a few alternative ways you can go about using native DOM methods:

1. Shadow DOM
2. Overriding all of them to put stuff where you want it

Obviously 1 is preferred over 2, but 2 is an option if you really want to keep the native DOM API without using native Shadow DOM. This is how a Shadow DOM polyfill would work underneath the hood anyway. For both of these, Skate allows you to do it, but does not do it for you.

## Scenario 1

```js
var myCustomEl = skate('my-custom-el', {
  template () {
    var sr = this.createShadowRoot();
    sr.innerHTML = '<div><content></content></div>';
  }
});

// <my-custom-el><div><p></p></div></my-custom-el>
myCustomEl().appendChild(document.createElement('p'));
```

## Scenario 2

```js
var myCustomEl = skate('my-custom-el', {
  prototype: {
    appendChild (node) {
      this.children[0].appendChild(node);
    }
  },
  template () {
    this.innerHTML = '<div></div>';
  }
});

// <my-custom-el><div><p></p></div></my-custom-el>
myCustomEl().appendChild(document.createElement('p'));
```

## Alternative Methods

I wouldn't really recommend any of the aforementioned methods, though, because the Shadow DOM spec is still under [heavy contention](https://hacks.mozilla.org/2015/06/the-state-of-web-components/) and overriding DOM methods can lead to pain.

### Properties (mutable state)

I would recommend using properties. It does change your public API, but it's very explicit.

```js
var myCustomEl = skate('my-custom-el', {
  prototype: {
    get paragraph () {
      let div = this.children[0];
      let p = div.children[0];
      return p;
    },
    set paragraph (p) {
      let div = this.children[0];
      div.innerHTML = '';
      div.appendChild(p);
    }
  },
  template () {
    this.innerHTML = '<div></div>';
  }
});

// <my-custom-el><div><p></p></div></my-custom-el>
myCustomEl().paragraph = document.createElement('p');
```

This can lead to a bit of boilerplate, but it's not hard to write abstractions around this. This is how [Shade's `content` binding](https://github.com/skatejs/shade#the-content-element) behaves, but it uses a [live collection](https://github.com/skatejs/shade#mutable-collection) to abstract the boilerplate.

### Properties (immutable state)

I recommend this option if you are able to diff your dom. We've started building a [DOM diffing abstraction](https://github.com/skatejs/dom-diff) in case you're interested in experimenting with it. For the sake of simplicity, I'll be using `innerHTML`.

```js
function render () {
  this.innerHTML = `<div>${this.content}</div>`;
}

var myCustomEl = skate('my-custom-el', {
  properties: {
    content: {
      set: render
    }
  },
  template: render
});

// <my-custom-el><div><p></p></div></my-custom-el>
myCustomEl().content = '<p></p>';
```

## Notes

I noticed you used `document.createElement()`. This is ok in native land, however, if you're running in browsers that don't support native custom elements, then you'll want to use something like `skate.create()` or the function / constructor returned from the `skate()` function call.

There's several reasons for this:

You don't have to worry about implementation details. For example, if you use `document.createElement()` you must remember how you've implemented the component in order to create it. For example, if you had the following components:

```js
skate('my-el-1', {});
skate('my-el-2', {
  extends: 'input'
});
```

In native you must know that you've extended an `input` for `my-el-2`:

```js
document.createElement('my-el-1');
document.createElement('input', 'my-el-2');
```

If you use `skate.create()`, you don't have to worry about that:

```js
// registering
skate.create('my-el-1');
skate.create('my-el-2');
```

Furthermore if you used the return value of `skate()`:

```js
let myEl2 = skate('my-el-2', {
  extends: 'input'
});
```

You could just call that like a function:

```js
myEl2();
```

These are going into the new documentation.

**Comment by antitoxic** [comment]
Amazing write up. Kudos.

**Comment by treshugart** [comment]
@antitoxic for posterity, in #285 it was discussed that we rename template to ready and move created before properties etc. This is being implemented by #307.

**Comment by antitoxic** [comment]
Best to comment on the pull?
